flang/include/flang/Optimizer/CodeGen/CGOps.td

```diff
@@ -82,6 +82,51 @@ def fircg_XEmboxOp : fircg_Op<"ext_embox", [AttrSizedOperandSegments]> {
   }];
 }
 
+// Extended rebox operation.
+def fircg_XReboxOp : fircg_Op<"ext_rebox", [AttrSizedOperandSegments]> {
+  let summary = "for internal conversion only";
+
+  let description = [{
+    Prior to lowering to LLVM IR dialect, a non-scalar non-trivial rebox op will
+    be converted to an extended rebox. This op will have the following sets of
+    arguments.
+
+       - box: The box being reboxed.
+       - newShape: A vector that is the new runtime shape for the array
+       - newShift: A vector that is the new runtime origin of the first element.
+         The default is a vector of the value 1.
+       - slice: A vector of triples that describe an array slice.
+       - subcomponent: A vector of indices for subobject slicing.
+       - sliceOrigins: A vector that is the new runtime origin of the slice operand
+
+    The box argument is mandatory, the other arguments are optional.
+  }];
+
+  let arguments = (ins
+    fir_BoxType:$box,
+    Variadic<AnyIntegerType>:$newShape,
+    Variadic<AnyIntegerType>:$newShift,
```

Comment:
Does `new` have significance? It seems redundant. The shape, shift, slice, and subcomps should all be applied on top of the array projection found in the originating box.

Reply:
It is to underline that the fir.box already has a shape that may not be the same. Distinguishing the old and new shape seemed important to me since the shape that is applied on the slice if any is the one from the fir.box, not the "newShape"/"newShift".
In this ops, we are manipulating 3 shapes, the one in the fir box, the one after the slicing, and the new shape that is given to the result. I am afraid "the shape" might be a bit ambiguous and I would rather be specific.

Reply:
Renamed `newXxx` to `xxx` after simplifying the rebox as discussed.